File: ruby/012.rb
def countFactors(n)
  count = 0

  i = 0
  while (i += 1)
    if n % i == 0 # i is a factor
      count += 1
      if i * i == n
        return count * 2 - 1
      elsif i * i > n
        return (count - 1) * 2
      end
    end

    if i * i > n
      return count * 2
    end
  end
end

goal = 500

triNum = 0

i = 0
while (i += 1)
  triNum += i
  if countFactors(triNum) > goal
    break
  end
end

puts triNum

File: r/012.r
#!/usr/bin/Rscript

countFactors <- function(n) {
    count <- 0
    i <- 0

    while (i <- i + 1) {
        if (n %% i == 0) { # i is a factor
            count <- count + 1
            if (i * i == n) {
                return(count * 2 - 1)
            } else if (i * i > n) {
                return((count - 1) * 2)
            }
        }
        if (i * i > n) {
            return(count * 2)
        }
    }
}

GOAL <- 500

triNum <- 0
i <- 0
while(i <- i + 1) {
    triNum <- triNum + i
    if (countFactors(triNum) > GOAL) { break }
}

cat(triNum, "\n")
